File: Transforme/domain/schoolUnity/SchoolUnity.fs
namespace Transforme.domain.schoolUnity

open System

type SchoolUnity() =
    let mutable _name: string option = None
    member val Id: Guid option = None with get, set

    member this.Name
        with get (): string option = _name
        and set (value) =
            match value with
            | Some str when String.IsNullOrWhiteSpace(str) ->
                invalidArg (nameof value) "Unidade escolar não pode ser nulo."
            | _ -> _name <- value

    static member Create(id: Guid, name: string) =
        let schoolUnity = SchoolUnity()

        schoolUnity.Id <- Some id
        schoolUnity.Name <- Some name
        schoolUnity

    override this.ToString() =
        let idStr =
            match this.Id with
            | Some guid -> guid.ToString("D")
            | None -> "N/A"

        let nameStr =
            match this.Name with
            | Some nome -> nome
            | None -> "N/A"

        $"SchoolUnity(Id = {idStr}, Nome = \"{nameStr}\")"

File: dotnet/backend/Tse_Backend.Domain/SchoolUnity.cs
namespace Tse_Backend.Domain;

public record SchoolUnity
{
    public SchoolUnity(Guid id, string name)
    {
        if (string.IsNullOrWhiteSpace(name)) throw new ArgumentException("School unity cannot be null.", nameof(name));

        Id = id;
        Name = name;
    }

    public Guid Id { get; }
    public string Name { get; }

    public static SchoolUnity Create(string name)
    {
        return new SchoolUnity(Guid.NewGuid(), name);
    }

    public static SchoolUnity Create(Guid id, string name)
    {
        return new SchoolUnity(id, name);
    }

    public override string ToString()
    {
        return $"SchoolUnity(Id = {Id:D}, Name = \"{Name}\")";
    }
}